compiler/testData/diagnostics/tests/smartCasts/variables/reassignedInRhs.fir.kt

```diff
@@ -0,0 +1,50 @@
+// SKIP_TXT
+
+import kotlin.contracts.*
+
+class C(val x: Int)
+
+@OptIn(ExperimentalContracts::class)
+fun isNotNullAlsoCall(a: String?, b: () -> Unit): Boolean {
+    contract {
+        returns(true) implies (a != null)
+        callsInPlace(b, InvocationKind.EXACTLY_ONCE)
+    }
+    b()
+    return a != null
+}
+
+fun binaryBooleanExpression() {
+    var x: String? = ""
+    if (x is String || (x is String).also { x = null }) {
+        x<!UNSAFE_CALL!>.<!>length // bad (x#0 is String, x#1 is Nothing?, this is either)
+    }
+}
+
+fun unoverriddenEquals(a: Any?) {
+    val c = C(1)
+    var b: Any?
+    b = c
+    if (b == c.also { b = a }) {
+        a.x // bad (b#0 is C, b#1 = a)
```

Comment:
I've created [KT-55096](https://youtrack.jetbrains.com/issue/KT-55096) to track this problem